cargo-shuttle/src/main.rs

```diff
@@ -3,143 +3,132 @@ mod client;
-use cargo_metadata::MetadataCommand;
-use shuttle_common::{ApiKey, project::ProjectConfig};
 use std::io::{Write};
 use std::{env, io};
 use std::fs::File;
-use std::path::Path;
 use std::rc::Rc;
 use structopt::StructOpt;
+use crate::config::RequestContext;
 
 #[tokio::main]
 async fn main() -> Result<()> {
-    let args: Args = Args::from_args();
-    match args {
-        Args::Deploy(deploy_args) => deploy(deploy_args).await,
-        Args::Status => status().await,
-        Args::Delete => delete().await,
-        Args::Auth(auth_args) => auth(auth_args).await,
-        Args::Login(login_args) => login(login_args).await,
-    }
+    Shuttle::new().run(Args::from_args()).await
 }
 
-async fn login(login_args: LoginArgs) -> Result<()> {
-    let api_key = login_args.api_key.unwrap_or_else(|| {
-        let url = "https://shuttle.rs/login";
+pub struct Shuttle {
+    ctx: RequestContext
+}
 
-        let _ = webbrowser::open(url);
+impl Shuttle {
+    pub fn new() -> Self {
+        let ctx = RequestContext::load_global().unwrap();
+        Self { ctx }
+    }
 
-        println!("If your browser did not automatically open, go to {url}");
-        print!("Enter Api Key: ");
-        
-        io::stdout().flush().unwrap();
+    pub async fn run(mut self, args: Args) -> Result<()> {
+        if matches!(args, Args::Deploy(..) | Args::Delete | Args::Status) {
+            self.load_project()?;
+        }
 
-        let mut input = String::new();
-        
-        io::stdin().read_line(&mut input).unwrap();    
+        match args {
+            Args::Deploy(deploy_args) => self.deploy(deploy_args).await,
+            Args::Status => self.status().await,
+            Args::Delete => self.delete().await,
+            Args::Auth(auth_args) => self.auth(auth_args).await,
+            Args::Login(login_args) => self.login(login_args).await,
+        }
+    }
 
-        input
-    });
+    pub fn load_project(&mut self) -> Result<()> {
+        let working_directory = env::current_dir()?;
+        self.ctx.load_local(working_directory)
+    }
 
-    config::create_with_api_key(api_key.trim().to_string())
-}
+    async fn login(&mut self, login_args: LoginArgs) -> Result<()> {
+        let api_key_str = login_args.api_key.unwrap_or_else(|| {
+            let url = "https://shuttle.rs/login";
 
-async fn auth(auth_args: AuthArgs) -> Result<()> {
-    if config::config_file_exists()? {
-        bail!("configuration file already exists")
-    }
-    let api_key = client::auth(auth_args.username)
-        .await
-        .context("failed to retrieve api key")?;
-    config::create_with_api_key(api_key)
-}
+            let _ = webbrowser::open(url);
 
-async fn delete() -> Result<()> {
-    let (api_key, project) = get_api_key_and_project()?;
-    client::delete(api_key, project)
-        .await
-        .context("failed to delete deployment")
-}
+            println!("If your browser did not automatically open, go to {url}");
```

Comment:
For some reason `{url}` keeps erroring on my system. Is the feature in rust stable yet?

Reply:
This is now working for me after using the Mozilla overlays for Rust